### DISCUSSION TITLE
Powershell / .NET in a Linux image

### DISCUSSION THREAD

**Discussion by geedsen** [PowerShell/PowerShell#20167] [Q&A]
I have been building an ubuntu image installing PowerShell on it. In different ways. First I had this 
![image](https://github.com/PowerShell/PowerShell/assets/1316595/1936e670-ffed-48e8-8762-b11ca40170f5)
But when we used this image to run a PowerShell command in our application, which is using PowerShellProcessInstance, we get an error that "-s" is an unknown cmdlet. "-s" we can see in the code is indeed automatically added. Is it because of the Powershell version we install?
I also noticed that in this case PowerShellProcessInstance.Process.StartInfo.FileName points to a "powershell.exe" not to "pwsh.exe". Setting it to "/usr/bin/pwsh" made it work (except for the -s" error).

I now created a different way of installing Powershell in our image. Using Powershell 7.4.0-preview.5. 
![image](https://github.com/PowerShell/PowerShell/assets/1316595/88c3eda0-ef16-4e0d-85ea-46c2f0f096b9)
Now the error is indeed gone. Is it because of the PowerShell version or because of the way I install it? 
What I still noticed is that PowerShellProcessInstance.Process.StartInfo.FileName now points to a pwsh.exe that does not exist. I still need to set it to "/usr/bin/pwsh".

What I also noticed it that the PowerShellProcessInstance.cs files contains the compiler switch LINUX. So how does this work? When do I get the package where it was compiled with the switch defined?  Is this when I actually build inside Linux and not on Windows?

**Comment by geedsen** [comment]
No help what so ever? Nobody knows when you get the "UNIX" version of the package?

**Reply by rhubarb-geek-nz** [reply]
I would expect the binaries built for rpm, deb and linux tar files will have been built for those platforms and cpus.

With C#/.NET there are two ways of doing cross platform, either dynamically by making decisions based on the current running host, or else with conditional macros for targeted builds.

Macros can be set up in conditional property groups in the csproj file when creating targeted builds. Also different compiler options will provide macros to indicate if the build is for a specific CPU or "any".

As PowerShell Core is designed for cross platform solutions I would expect PowerShell user code in assemblies to avoid platform specific builds unless absolutely necessary, as you then have to carefully package multiple binaries in PowerShell modules.

**Comment by rhubarb-geek-nz** [comment]
The symbolic link is `/usr/bin/pwsh-preview` rather than` /usr/bin/pwsh`

```
$ dpkg-deb -c powershell-preview_7.4.0-preview.5-1.deb_amd64.deb | grep pwsh
-rw-r--r-- 0/0             485 2023-08-17 05:40 ./usr/local/share/man/man1/pwsh-preview.1.gz
lrwxrwxrwx 0/0               0 2023-08-17 05:40 ./usr/bin/pwsh-preview -> /opt/microsoft/powershell/7-preview/pwsh
-rw-r--r-- 0/0            8672 2023-08-17 04:54 ./opt/microsoft/powershell/7-preview/pwsh.xml
-rw-r--r-- 0/0             582 2023-08-17 04:54 ./opt/microsoft/powershell/7-preview/pwsh.runtimeconfig.json
-rwxr-xr-x 0/0           73672 2023-08-17 04:54 ./opt/microsoft/powershell/7-preview/pwsh
-rw-r--r-- 0/0          130773 2023-08-17 04:55 ./opt/microsoft/powershell/7-preview/pwsh.deps.json
-rw-r--r-- 0/0          149032 2023-08-17 05:27 ./opt/microsoft/powershell/7-preview/pwsh.dll
```
